File: HW/HW10/TreeNode.java
import java.util.*;

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

 class LevelOrderTraversal {
    public static void main(String[] args) {
        TreeNode root = new TreeNode(4);
        root.left = new TreeNode(3);
        root.right = new TreeNode(8);
        root.left.left = new TreeNode(1);
        root.right.left = new TreeNode(5);
        root.right.right = new TreeNode(9);
        LevelOrderTraversal ob = new LevelOrderTraversal();

        ob.printLevelTraversal(ob.TreeLevelByLevel(root));
    }

    public void printLevelTraversal(List<List<TreeNode>> totalTreeLevel) {
        if (totalTreeLevel != null) {
            int counter_1 = 0;
            int counter_2 = 0;
            System.out.print("[");
            for (List<TreeNode> levelCountNodes : totalTreeLevel) {
                System.out.print("[");
                for (TreeNode node : levelCountNodes) {
                    if (counter_2 < levelCountNodes.size() - 1)
                        System.out.print(node.val + ", ");
                    else if (counter_1 < totalTreeLevel.size() - 1)
                        System.out.print(node.val + "],");
                    else
                        System.out.print(node.val + "]");
                    counter_2++;

                }
                counter_1++;
                counter_2 = 0;
            }
            System.out.println("]");

        } else
            System.out.println("The tree is empty");

    }

    public List<List<TreeNode>> TreeLevelByLevel(TreeNode root) {
        if (checkCount(root, 0) < 0 || checkCount(root, 0) > 2000 || !checkNodeVal(root))
            throw new IllegalArgumentException("The conditions are not met");

        if (root == null) {
            return null;
        }

        List<List<TreeNode>> totalTreeLevel = new ArrayList<>();
        List<TreeNode> currentTreeLevel = new ArrayList<>();
        currentTreeLevel.add(root);

        while (!currentTreeLevel.isEmpty()) {
            List<TreeNode> nextTreeLevel = new ArrayList<>();
            List<TreeNode> currentLevelValues = new ArrayList<>();
            for (TreeNode node : currentTreeLevel) {
                currentLevelValues.add(node);
                if (node.left != null) {
                    nextTreeLevel.add(node.left);
                }
                if (node.right != null) {
                    nextTreeLevel.add(node.right);
                }
            }
            totalTreeLevel.add(currentLevelValues);
            currentTreeLevel = nextTreeLevel;
        }
        return totalTreeLevel;
    }

    public int checkCount(TreeNode node, int countNode) {
        if (node == null)
            return countNode;

        countNode++;
        countNode = checkCount(node.left, countNode);
        countNode = checkCount(node.right, countNode);
        return countNode;
    }

    public boolean checkNodeVal(TreeNode node) {
        if (node == null)
            return true;

        if (node.val < -1000 || node.val > 1000)
            return false;

        return checkNodeVal(node.left) && checkNodeVal(node.right);

    }

}

File: HW/HW10/TreeNode.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class LevelOrderTraversal:
    def __init__(self):
        pass

    def printLevelTraversal(self, totalTreeLevel):
        if totalTreeLevel:
            counter_1 = 0
            counter_2 = 0
            print("[", end="")
            for levelCountNodes in totalTreeLevel:
                print("[", end="")
                for node in levelCountNodes:
                    if counter_2 < len(levelCountNodes) - 1:
                        print(node.val, end=", ")
                    elif counter_1 < len(totalTreeLevel) - 1:
                        print(node.val, end="],")
                    else:
                        print(node.val, end="]")
                    counter_2 += 1

                counter_1 += 1
                counter_2 = 0
            print("]")

        else:
            print("The tree is empty")

    def treeLevelByLevel(self, root):
        if self.checkCount(root, 0) < 0 or self.checkCount(root, 0) > 2000 or not self.checkNodeVal(root):
            raise ValueError("The conditions are not met")

        if not root:
            return None

        totalTreeLevel = []
        currentTreeLevel = [root]

        while currentTreeLevel:
            nextTreeLevel = []
            currentLevelValues = []
            for node in currentTreeLevel:
                currentLevelValues.append(node)
                if node.left:
                    nextTreeLevel.append(node.left)
                if node.right:
                    nextTreeLevel.append(node.right)
            totalTreeLevel.append(currentLevelValues)
            currentTreeLevel = nextTreeLevel

        return totalTreeLevel

    def checkCount(self, node, countNode):
        if not node:
            return countNode

        countNode += 1
        countNode = self.checkCount(node.left, countNode)
        countNode = self.checkCount(node.right, countNode)
        return countNode

    def checkNodeVal(self, node):
        if not node:
            return True

        if node.val < -1000 or node.val > 1000:
            return False

        return self.checkNodeVal(node.left) and self.checkNodeVal(node.right)

    def main(self):
        # Creating the sample binary tree
        root = TreeNode(4)
        root.left = TreeNode(3)
        root.right = TreeNode(8)
        root.left.left = TreeNode(1)
        root.right.left = TreeNode(5)
        root.right.right = TreeNode(9)

        # Performing level-order traversal
        totalTreeLevel = self.treeLevelByLevel(root)

        # Printing the result
        self.printLevelTraversal(totalTreeLevel)

if __name__ == "__main__":
    # Create an instance of LevelOrderTraversal and call the main function
    ob = LevelOrderTraversal()
    ob.main()

